src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -83,43 +79,57 @@
   flush-pool!
   "Flush of the current JRuby pool.  NOTE: this function should never
   be called except by the pool-agent."
-  [pool-context :- jruby-schemas/PoolContext]
+  [pool-context :- jruby-schemas/PoolContext
+   reason :- jruby-schemas/FlushReason
+   flush-complete? :- IDeref]
   ;; Since this function is only called by the pool-agent, and since this
   ;; is the only entry point into the pool flushing code that is exposed by the
   ;; service API, we know that if we receive multiple flush requests before the
   ;; first one finishes, they will be queued up and the body of this function
   ;; will be executed atomically; we don't need to worry about race conditions
   ;; between the steps we perform here in the body.
-  (log/info "Flush request received; creating new JRuby pool.")
-  (let [{:keys [config profiler pool-state]} pool-context
-        new-pool-state (jruby-internal/create-pool-from-config config)
-        new-pool (:pool new-pool-state)
-        old-pool @pool-state
-        count    (:size old-pool)]
-    (log/info "Replacing old JRuby pool with new instance.")
-    (reset! pool-state new-pool-state)
-    (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range count)]
-      (try
-        (let [id        (inc i)
-              instance  (jruby-internal/borrow-from-pool!*
-                          jruby-internal/borrow-without-timeout-fn
-                          (:pool old-pool))]
-          (try
-            (flush-instance! pool-context instance new-pool id config profiler)
-            (finally
-              (.releaseItem (:pool old-pool) instance false)))
-          (log/infof "Finished creating JRubyPuppet instance %d of %d"
-                     id count))
-        (catch Exception e
-          (.clear new-pool)
-          (.insertPill new-pool (PoisonPill. e))
-          (throw (IllegalStateException.
-                   "There was a problem adding a JRubyPuppet instance to the pool."
-                   e)))))
-    ;; Add a "RetryPoisonPill" to the pool in case something else is in the
-    ;; process of borrowing from the old pool.
-    (.insertPill (:pool old-pool) (RetryPoisonPill. (:pool old-pool)))))
+  (try
+    (log/info "Flush request received; creating new JRuby pool.")
+    (let [{:keys [config profiler pool-state]} pool-context
+          new-pool-state (jruby-internal/create-pool-from-config config)
+          new-pool (:pool new-pool-state)
+          old-pool @pool-state
+          old-pool-size (:size old-pool)]
+      (log/info "Replacing old JRuby pool with new instance.")
+      (if (and (= reason :shutdown) (not= old-pool-size (count (.getRegisteredElements (:pool old-pool)))))
+        (log/warn "Attempting to flush a pool that does not appear to have successfully initialized. Aborting.")
```

Comment:
The only time @cprice404 and I saw this was in a test that replaced the pool creation function with an exception being thrown. I'm not sure what the right thing to do in that case is, as the pool creation exception will have already started a shutdown.